### DISCUSSION TITLE
use reflection In flutter

### DISCUSSION THREAD

**Discussion by chamodyauthpalabandara** [google/reflectable.dart#333] [Q&A]
In the Flutter application, I use the interface class to implement the methods and in the main class I need to find dynamically classes that were implemented using the ISyncService interface and execute all of them 

here is an example of a classes

abstract class ISyncService {
  Future<void> fetchAndSyncData();
}

class InvoiceSyncService implements ISyncService {
  @override
  Future<void> fetchAndSyncData() async {
    // Implementation for syncing invoices
    print('Syncing invoices...');
    // Perform sync logic here
  }
}

class ProductSyncService implements ISyncService {
  @override
  Future<void> fetchAndSyncData() async {
    // Implementation for syncing products
    print('Syncing products...');
    // Perform sync logic here
  }
}

Can I use reflection to achieve this, like I do with other languages (.net, java)?
if it can't can you suggest some method to achieve this task?

**Comment by eernstg** [comment]
You can certainly do that. Here's an example:

```dart
// ----- Library 'main.dart'.

import 'package:reflectable/reflectable.dart';
import 'main.reflectable.dart';

class Reflector extends Reflectable {
  const Reflector() : super(subtypeQuantifyCapability, newInstanceCapability);
}

const reflector = Reflector();

@reflector
abstract class ISyncService {
  Future fetchAndSyncData();
}

void main() {
  initializeReflectable();
  for (var classMirror in reflector.annotatedClasses) {
    if (classMirror.simpleName != 'ISyncService') {
      dynamic instance = classMirror.newInstance('', const [], const {});
      if (instance is ISyncService) {
        instance.fetchAndSyncData();
      }
    }
  }
}

// ------------------------------------------------------------
// The classes below could be in other libraries. There is no
// need for 'main.dart' to contain a complete list of subtypes
// of `ISyncService` in any way, they will be included in 
// `reflector.annotatedClasses` because `reflector` has the 
// capability `subtypeQuantifyCapability`.

class InvoiceSyncService implements ISyncService {
  @override
  Future<void> fetchAndSyncData() async {
    // Implementation for syncing invoices
    print('Syncing invoices...');
    // Perform sync logic here
  }
}

class ProductSyncService implements ISyncService {
  @override
  Future<void> fetchAndSyncData() async {
    // Implementation for syncing products
    print('Syncing products...');
    // Perform sync logic here
  }
}
```

Note that reflectable is based on code generation, and you need to perform the code generation starting from the entry point of the application that you want to cover (that is, the library that contains the `main` function). Let's say that the above is stored in `bin/main.dart` in a package that contains your app. Then, from the root of your package:

```
> dart pub get
> dart run build_runner build bin
> dart bin/main.dart
```

Things will be slightly different when in Flutter, but I used plain Dart in order to make the example as simple as possible.
